custom/enikshay/integrations/nikshay/tests/test_repeaters.py

```diff
@@ -593,3 +598,316 @@ def test_payload_properties(self, _):
+        # skip if episode case not nikshay registered
+        self.update_case_with(self.test_id, {"date_tested": datetime.now()})
+        self.assertFalse(check_repeat_record_added())
+
+        self.update_case_with(self.episode_id, {"nikshay_registered": 'true'})
+
+        # skip if episode case has no nikshay_id
+        self.update_case_with(self.test_id, {"date_tested": datetime.now()})
+        self.assertFalse(check_repeat_record_added())
+
+        self.update_case_with(self.episode_id, {"nikshay_id": DUMMY_NIKSHAY_ID})
+
+        self.update_case_with(self.test_id, {"date_tested": datetime.now()})
+        self.assertTrue(check_repeat_record_added())
+
+        # skip if test submission
+        self.dmc.metadata['is_test'] = 'yes'
+        self.dmc.save()
+        self.update_case_with(self.test_id, {"date_tested": datetime.now()})
+        self.assertFalse(check_repeat_record_added())
+        self.dmc.metadata['is_test'] = 'no'
+        self.dmc.save()
+
+        self.update_case_with(self.test_id, {"date_tested": datetime.now()})
+        self.assertTrue(check_repeat_record_added())
+
+        # allow update for diagnostic tests irrespective of the follow up test reason
+        self.update_case_with(self.test_id, {
+            "date_tested": datetime.now(),
+            "purpose_of_testing": 'diagnostic',
+            "follow_up_test_reason": 'end_of_no_p',
+        })
+        self.assertTrue(check_repeat_record_added())
+
+        # ensure followup test reason is in the allowed ones
+        self.update_case_with(self.test_id, {
+            "date_tested": datetime.now(),
+            "purpose_of_testing": 'just like that',
+            "follow_up_test_reason": 'end_of_no_p'
+        })
+        self.assertFalse(check_repeat_record_added())
+
+        self.update_case_with(self.test_id, {
+            "date_tested": datetime.now(),
+            "purpose_of_testing": 'just like that',
+            "follow_up_test_reason": 'end_of_ip',
+        })
+        self.assertTrue(check_repeat_record_added())
+
+        # ensure date tested is added for the test
+        self.update_case_with(self.test_id, {
+            "follow_up_test_reason": 'end_of_ip'
+        })
+        self.assertFalse(check_repeat_record_added())
+
+        # ensure test type is in the allowed ones
+        self.update_case_with(self.test_id, {
+            "date_tested": datetime.now(),
+            "follow_up_test_reason": 'end_of_ip',
+            "test_type_value": 'irrelevant-test'
+        })
+        self.assertFalse(check_repeat_record_added())
+
+
+class TestNikshayFollowupPayloadGenerator(ENikshayLocationStructureMixin, NikshayRepeaterTestBase):
+    def setUp(self):
+        super(TestNikshayFollowupPayloadGenerator, self).setUp()
+
+        self.cases = self.create_case_structure()
+        self.test_case = self.cases['test']
+
+        self._create_nikshay_registered_case()
+
+        MockRepeater = namedtuple('MockRepeater', 'username password')
```

Comment:
looks like this is copy/pasted from other tests. might be worth making it part of the super class?